src/slurm_plugin/clustermgtd.py

```diff
@@ -904,6 +1042,135 @@ def _terminate_orphaned_instances(self, cluster_instances, ips_used_by_slurm):
+        """Retrieve ACTIVE partitions that have bootstrap failures reach max bootstrap failures."""
+        bootstrap_failure_partitions = []
+        for p in active_partitions_protected_failure_count_map:
+            if active_partitions_protected_failure_count_map.get(p) >= self._config.protected_failure_count:
+                bootstrap_failure_partitions.append(p)
+        if bootstrap_failure_partitions:
+            log.info(f"Bootstrap failure partitions: {bootstrap_failure_partitions}")
+        return bootstrap_failure_partitions
+
+    @staticmethod
+    def _get_partitions_have_active_jobs(slurm_nodes):
+        """Retrieve partitions currently have active jobs running."""
+        partitions_have_active_jobs = []
+        for node in slurm_nodes:
+            if node.is_running_job():
+                partitions_have_active_jobs += node.partitions
+        return set(partitions_have_active_jobs)
+
+    def _handle_health_check_bootsrap_failure_node(self, unhealthy_node):
+        """Increase the partition failure count and add failed node to the set of bootstrap failure nodes."""
+        self._increase_partitions_protected_failure_count(unhealthy_node)
+        # Add the node type to bootstrap failure nodes types
+        self._bootstrap_failure_nodes_types.add(get_node_type(unhealthy_node.name))
+
+    def _get_bootstrap_failure_nodes_types_successfully_launched(self, slurm_nodes, private_ip_to_instance_map):
+        """Find nodes types which have been failed during bootstrap now become online."""
+        online_nodes = self._find_online_slurm_nodes(slurm_nodes, private_ip_to_instance_map)
+        online_nodes_types = get_nodes_types_from_slurm_nodes(online_nodes)
+
+        bootstrap_failure_nodes_types_successfully_launched = online_nodes_types.intersection(
+            self._bootstrap_failure_nodes_types
+        )
+        return bootstrap_failure_nodes_types_successfully_launched
+
+    def _get_partitions_to_disable_and_bootstrap_failure_partitions_have_jobs(
+        self, slurm_nodes, active_partitions_protected_failure_count_map
+    ):
+        """Get partitions not running jobs to disable and get bootstrap failure partitions have job running."""
+        # Get partitions have jobs running
+        partitions_have_active_jobs = ClusterManager._get_partitions_have_active_jobs(slurm_nodes)
+        # Get bootstrap failure partitions
+        bootstrap_failure_partitions = self._get_bootstrap_failure_partitions(
+            active_partitions_protected_failure_count_map
+        )
+        # Only disable bootstrap failure partitions that not having active jobs running
+        partitions_to_disable = set(bootstrap_failure_partitions) - partitions_have_active_jobs
+        bootstrap_failure_partitions_have_jobs = set(bootstrap_failure_partitions) - partitions_to_disable
+        return partitions_to_disable, bootstrap_failure_partitions_have_jobs
```

Comment:
We only use `bootstrap_failure_partitions_have_jobs` to print out a log message. Can we just print out a message here and avoid passing around anything we don't really need?